core/src/main/java/org/apache/iceberg/MetricsUtil.java

```diff
@@ -56,4 +64,123 @@ public static MetricsModes.MetricsMode metricsMode(
+   * @return map of readable column name to column metric, of which the bounds are made readable
+   */
+  public static Map<String, StructLike> readableMetricsMap(
+      Schema schema, Map<Integer, String> namesById, ContentFile<?> contentFile) {
+    Map<String, StructLike> metricsStruct = Maps.newHashMapWithExpectedSize(namesById.size());
+
+    Map<Integer, Long> columnSizes = contentFile.columnSizes();
+    Map<Integer, Long> valueCounts = contentFile.valueCounts();
+    Map<Integer, Long> nullValueCounts = contentFile.nullValueCounts();
+    Map<Integer, Long> nanValueCounts = contentFile.nanValueCounts();
+    Map<Integer, ByteBuffer> lowerBounds = contentFile.lowerBounds();
+    Map<Integer, ByteBuffer> upperBounds = contentFile.upperBounds();
+
+    for (int id : namesById.keySet()) {
+      Types.NestedField field = schema.findField(id);
+      if (field.type().isPrimitiveType()) {
+        // Iceberg stores metrics only for primitive types
+        String colName = namesById.get(id);
+        ReadableMetricsStruct struct =
+            new ReadableMetricsStruct(
+                columnSizes == null ? null : columnSizes.get(id),
+                valueCounts == null ? null : valueCounts.get(id),
+                nullValueCounts == null ? null : nullValueCounts.get(id),
+                nanValueCounts == null ? null : nanValueCounts.get(id),
+                lowerBounds == null ? null : convertToReadable(field, lowerBounds.get(id)),
+                upperBounds == null ? null : convertToReadable(field, upperBounds.get(id)));
+        metricsStruct.put(colName, struct);
+      }
+    }
+    return metricsStruct;
+  }
+
+  public static String convertToReadable(Types.NestedField field, ByteBuffer value) {
+    if (field == null || value == null) {
+      return null;
+    }
+    try {
+      return Transforms.identity(field.type())
+          .toHumanString(Conversions.fromByteBuffer(field.type(), value));
+    } catch (Exception e) {
+      LOG.warn("Error converting metric to readable form", e);
+      return null;
+    }
+  }
+
+  public static class ReadableMetricsStruct implements StructLike {
+
+    private final Long columnSize;
+    private final Long valueCount;
+    private final Long nullValueCount;
+    private final Long nanValueCount;
+    private final String lowerBound;
+    private final String upperBound;
+
+    public ReadableMetricsStruct(
+        Long columnSize,
+        Long valueCount,
+        Long nullValueCount,
+        Long nanValueCount,
+        String lowerBound,
+        String upperBound) {
+      this.columnSize = columnSize;
+      this.valueCount = valueCount;
+      this.nullValueCount = nullValueCount;
+      this.nanValueCount = nanValueCount;
+      this.lowerBound = lowerBound;
+      this.upperBound = upperBound;
+    }
+
+    @Override
+    public int size() {
+      return 6;
+    }
+
+    @Override
+    public <T> T get(int pos, Class<T> javaClass) {
+      Object value;
+      switch (pos) {
```

Comment:
nit: Would having `private Object get(int pos)` help separate getting the value from casting?

Reply:
Done